### PR TITLE
feat: more node schema, ops

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

**Comment by rphmeier** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/14?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/15?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/14?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#13** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/13?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: core/src/lib.rs [hunk 1]
```diff
@@ -25,4 +39,72 @@ pub fn is_internal(hash: &NodeHash) -> bool {
...
+}
+
+/// A trie node hash function specialized for 64 bytes of data.
+pub trait NodeHasher {
+    /// Hash a node, encoded as exactly 64 bytes of data. This should not
+    /// domain-separate the hash.
+    fn hash_node(data: &[u8; 64]) -> NodeHash;
+}
+
+pub trait NodeHasherExt: NodeHasher {
```
Comment (pepyakin) [inline core/src/lib.rs:90]: Not sure what purpose does this ext serve.

File: core/src/lib.rs [hunk 2]
```diff
@@ -9,7 +9,21 @@
...
+pub type ValueHash = [u8; 32];
+
+/// The encoded state of a node. In this schema, it is always 512 bits.
+///
+/// Note that this encoding itself does not contain information about
+/// whether the node is a leaf or internal node.
+pub type NodeEncoding = [u8; 64];
+
+/// Denotes an empty sub-tree beyond this point. Concretely, when this appears within
+/// an internal node, it implies that no keys beginning with the path to the internal node plus the child's bit
```
Comment (pepyakin) [inline core/src/lib.rs:25]: Would be great if we agreed to stick to 100 soft and 120 hard wrap guidelines.

**Review by pepyakin — APPROVED** [review]

**Comment by pepyakin** [comment]
### Merge activity

* **Mar 21, 11:56 AM EDT**: @pepyakin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/14).

**Pull request merged**